File: backend/app/api/colab.py
"""
API endpoints for Google Colab integration.
"""
from fastapi import APIRouter, HTTPException, UploadFile, File
from pydantic import BaseModel, Field
from typing import Optional, Dict, Any
import json

from app.services.colab_service import ColabService

router = APIRouter()


class ExportRequest(BaseModel):
    """Request model for exporting code to notebook."""
    code: str = Field(..., description="Python code to export")
    metadata: Optional[Dict[str, Any]] = Field(
        None,
        description="Optional metadata (exercise_id, module_id, etc.)"
    )


class ExportResponse(BaseModel):
    """Response model for export endpoint."""
    success: bool
    notebook: Dict[str, Any]
    message: str


class ImportRequest(BaseModel):
    """Request model for importing from Colab URL or notebook data."""
    url: Optional[str] = Field(None, description="Google Colab URL")
    notebook: Optional[Dict[str, Any]] = Field(None, description="Notebook JSON data")


class ImportResponse(BaseModel):
    """Response model for import endpoint."""
    success: bool
    code: str
    metadata: Dict[str, Any]
    validation: Dict[str, Any]
    message: str


@router.post("/export", response_model=ExportResponse)
async def export_to_colab(request: ExportRequest):
    """
    Export Python code to Jupyter notebook format (.ipynb).
    
    This endpoint converts Python code from the code editor into a valid
    Jupyter notebook format that can be opened in Google Colab.
    
    - **code**: Python code to export
    - **metadata**: Optional metadata to include in the notebook
    
    Returns the notebook as a JSON structure that can be saved as .ipynb
    """
    try:
        colab_service = ColabService()
        notebook = colab_service.export_to_notebook(
            code=request.code,
            metadata=request.metadata
        )
        
        return ExportResponse(
            success=True,
            notebook=notebook,
            message="Notebook generado exitosamente"
        )
    
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Error al exportar notebook: {str(e)}"
        )


@router.post("/import", response_model=ImportResponse)
async def import_from_colab(request: ImportRequest):
    """
    Import code from a Google Colab notebook.
    
    This endpoint accepts either:
    - A Google Colab URL (will be validated but not fetched - client handles download)
    - A notebook JSON structure directly
    
    The notebook is parsed and code is extracted from code cells.
    Markdown cells are converted to Python comments.
    
    - **url**: Google Colab URL (optional)
    - **notebook**: Notebook JSON data (optional)
    
    Returns the extracted Python code and validation results
    """
    try:
        colab_service = ColabService()
        
        # Validate that at least one input is provided
        if not request.url and not request.notebook:
            raise HTTPException(
                status_code=400,
                detail="Debe proporcionar una URL de Colab o datos de notebook"
            )
        
        # If URL is provided, validate it
        if request.url:
            if not colab_service.validate_colab_url(request.url):
                raise HTTPException(
                    status_code=400,
                    detail="URL de Google Colab inválida. Debe ser una URL de colab.research.google.com"
                )
            
            # Note: In a real implementation, you might want to fetch the notebook
            # from the URL here. For now, we expect the client to handle this.
            if not request.notebook:
                return ImportResponse(
                    success=False,
                    code="",
                    metadata={},
                    validation={"valid": False},
                    message="URL válida. Por favor, descargue el notebook y envíelo en el campo 'notebook'"
                )
        
        # Import from notebook data
        if not request.notebook:
            raise HTTPException(
                status_code=400,
                detail="Datos de notebook requeridos"
            )
        
        result = colab_service.import_from_notebook(request.notebook)
        code = result['code']
        metadata = result['metadata']
        
        # Validate syntax
        validation = colab_service.validate_notebook_syntax(code)
        
        if not validation['valid']:
            return ImportResponse(
                success=False,
                code=code,
                metadata=metadata,
                validation=validation,
                message=f"El código importado contiene errores de sintaxis: {validation.get('error', 'Error desconocido')}"
            )
        
        return ImportResponse(
            success=True,
            code=code,
            metadata=metadata,
            validation=validation,
            message="Código importado exitosamente desde Google Colab"
        )
    
    except ValueError as e:
        raise HTTPException(
            status_code=400,
            detail=str(e)
        )
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Error al importar notebook: {str(e)}"
        )


@router.post("/import/file")
async def import_from_file(file: UploadFile = File(...)):
    """
    Import code from an uploaded .ipynb file.
    
    This endpoint accepts a file upload of a Jupyter notebook (.ipynb)
    and extracts the Python code from it.
    
    - **file**: .ipynb file to import
    
    Returns the extracted Python code and validation results
    """
    try:
        # Validate file extension
        if not file.filename.endswith('.ipynb'):
            raise HTTPException(
                status_code=400,
                detail="El archivo debe tener extensión .ipynb"
            )
        
        # Read and parse file
        content = await file.read()
        try:
            notebook_data = json.loads(content)
        except json.JSONDecodeError:
            raise HTTPException(
                status_code=400,
                detail="El archivo no es un JSON válido"
            )
        
        # Import using the service
        colab_service = ColabService()
        result = colab_service.import_from_notebook(notebook_data)
        code = result['code']
        metadata = result['metadata']
        
        # Validate syntax
        validation = colab_service.validate_notebook_syntax(code)
        
        if not validation['valid']:
            return ImportResponse(
                success=False,
                code=code,
                metadata=metadata,
                validation=validation,
                message=f"El código importado contiene errores de sintaxis: {validation.get('error', 'Error desconocido')}"
            )
        
        return ImportResponse(
            success=True,
            code=code,
            metadata=metadata,
            validation=validation,
            message="Código importado exitosamente desde archivo"
        )
    
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=500,
            detail=f"Error al procesar archivo: {str(e)}"
        )


@router.post("/validate-url")
async def validate_colab_url(url: str):
    """
    Validate a Google Colab URL.
    
    - **url**: URL to validate
    
    Returns validation result
    """
    colab_service = ColabService()
    is_valid = colab_service.validate_colab_url(url)
    
    return {
        "valid": is_valid,
        "message": "URL válida de Google Colab" if is_valid else "URL inválida. Debe ser una URL de colab.research.google.com"
    }

File: frontend/src/lib/api/colab.ts
/**
 * API client for Google Colab integration
 */

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000';

export interface ExportRequest {
  code: string;
  metadata?: {
    exercise_id?: number;
    module_id?: number;
    lesson_id?: number;
    user_id?: number;
    [key: string]: any;
  };
}

export interface ExportResponse {
  success: boolean;
  notebook: any;
  message: string;
}

export interface ImportRequest {
  url?: string;
  notebook?: any;
}

export interface ImportResponse {
  success: boolean;
  code: string;
  metadata: Record<string, any>;
  validation: {
    valid: boolean;
    error?: string;
    line?: number;
    offset?: number;
  };
  message: string;
}

/**
 * Export code to Jupyter notebook format
 */
export async function exportToColab(
  code: string,
  metadata?: ExportRequest['metadata']
): Promise<ExportResponse> {
  const response = await fetch(`${API_URL}/api/colab/export`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ code, metadata }),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Error al exportar notebook');
  }

  return response.json();
}

/**
 * Import code from Colab notebook
 */
export async function importFromColab(
  request: ImportRequest
): Promise<ImportResponse> {
  const response = await fetch(`${API_URL}/api/colab/import`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(request),
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Error al importar notebook');
  }

  return response.json();
}

/**
 * Import code from uploaded .ipynb file
 */
export async function importFromFile(file: File): Promise<ImportResponse> {
  const formData = new FormData();
  formData.append('file', file);

  const response = await fetch(`${API_URL}/api/colab/import/file`, {
    method: 'POST',
    body: formData,
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Error al importar archivo');
  }

  return response.json();
}

/**
 * Validate a Google Colab URL
 */
export async function validateColabUrl(url: string): Promise<{ valid: boolean; message: string }> {
  const response = await fetch(`${API_URL}/api/colab/validate-url?url=${encodeURIComponent(url)}`, {
    method: 'POST',
  });

  if (!response.ok) {
    const error = await response.json();
    throw new Error(error.detail || 'Error al validar URL');
  }

  return response.json();
}

/**
 * Download notebook as .ipynb file
 */
export function downloadNotebook(notebook: any, filename: string = 'notebook.ipynb') {
  const blob = new Blob([JSON.stringify(notebook, null, 2)], {
    type: 'application/json',
  });
  const url = URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = filename;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  URL.revokeObjectURL(url);
}

/**
 * Open notebook in Google Colab
 * This creates a temporary file and opens it in Colab
 */
export function openInColab(notebook: any) {
  // Convert notebook to JSON string
  const notebookJson = JSON.stringify(notebook, null, 2);
  
  // Create a blob and download it
  const blob = new Blob([notebookJson], { type: 'application/json' });
  const url = URL.createObjectURL(blob);
  
  // Create temporary download link
  const link = document.createElement('a');
  link.href = url;
  link.download = 'plataforma_educativa_notebook.ipynb';
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  URL.revokeObjectURL(url);
  
  // Open Colab in new tab with instructions
  const colabUrl = 'https://colab.research.google.com/';
  window.open(colabUrl, '_blank');
  
  // Show instructions to user
  return {
    success: true,
    message: 'Notebook descargado. Por favor, sube el archivo en Google Colab usando "Archivo > Subir notebook"'
  };
}
